### ISSUE TITLE
Software Design: Introduce nonce caching

### ISSUE THREAD

**Issue by haerdib** [scs/substrate-api-client#318]
Currently, every `compose_extrinsic!` call, calls the substrate storage. There's already a workaround for this by using `ccompose_extrinsic_offline!`, but it sure would be nice, to have a nonce cache that caches the nonce, and updates it with a storage request in case of a failure (e.g. mark it as dirty or something). 

Complexity: When to revert nonce? What about future?

**Comment by Niederb** [comment]
Maybe in a first version we can introduce a `NonceProvider` trait that abstracts over getting the nonce and provide two simple implementations:
- `LocalNonceProvider`: Fetches the nonce only once and increases it automatically, but does not implement additional logic (fetching nonce again on error, ...)
- `OnlineNonceProvider`: The approach currently used that always fetches the nonce from the node

This will already be some work to think about how the trait should look, how to configure this and how to pass the `NonceProvider` around. We could then implement fancier nonce caching later on.

**Comment by haerdib** [comment]
I agree in general, but before doing all this work, we should probably make sure that a local nonce cache does make sense. At least theoretically. E.g. find out if such a cache could ever be robust enough to be applied in the field. What do you think?

**Comment by Niederb** [comment]
Yes, I agree and I think even this "simple" approach is more involved that it seems at first glance.
